File: blind75/75_graph_valid_tree/valid_tree.py
from typing import List


def valid_tree(n: int, edges: List[List[int]]) -> bool:
    """
    Checks if the edges form a valid tree using DFS.
    Time: O(E + n), Space: O(n)

    Args:
        n: Number of nodes
        edges: List of edges [u, v]

    Returns:
        True if edges form a valid tree, False otherwise
    """
    # A tree with n nodes must have exactly n-1 edges
    if len(edges) != n - 1:
        return False

    # Build adjacency list
    graph = {i: [] for i in range(n)}
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)

    # Check if all nodes are connected using DFS
    visited = set()

    def dfs(node):
        visited.add(node)
        for neighbor in graph[node]:
            if neighbor not in visited:
                dfs(neighbor)

    dfs(0)
    return len(visited) == n


class UnionFind:
    """Union-Find (Disjoint Set Union) data structure."""

    def __init__(self, n: int):
        """Initialize Union-Find with n disjoint sets."""
        self.parent = list(range(n))
        self.rank = [1] * n

    def find(self, x: int) -> int:
        """Find the root of the set containing x with path compression."""
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        """
        Union the sets containing x and y.
        Returns False if they're already in the same set (cycle detected).
        """
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x == root_y:
            return False  # Already in same set - cycle detected

        # Union by rank
        if self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
        elif self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += 1

        return True


def valid_tree_union_find(n: int, edges: List[List[int]]) -> bool:
    """
    Checks if the edges form a valid tree using Union-Find.
    Time: O(E * α(n)) ≈ O(E), Space: O(n)

    Args:
        n: Number of nodes
        edges: List of edges [u, v]

    Returns:
        True if edges form a valid tree, False otherwise
    """
    # A tree with n nodes must have exactly n-1 edges
    if len(edges) != n - 1:
        return False

    uf = UnionFind(n)

    for u, v in edges:
        # If union returns False, we found a cycle
        if not uf.union(u, v):
            return False

    # If we have n-1 edges and no cycles, it's a valid tree
    return True

File: blind75/75_graph_valid_tree/valid_tree.go
package valid_tree

// ValidTree checks if the edges form a valid tree using DFS.
// Time: O(E + n), Space: O(n)
func ValidTree(n int, edges [][]int) bool {
	// A tree with n nodes must have exactly n-1 edges
	if len(edges) != n-1 {
		return false
	}

	// Build adjacency list
	graph := make(map[int][]int)
	for _, edge := range edges {
		u, v := edge[0], edge[1]
		graph[u] = append(graph[u], v)
		graph[v] = append(graph[v], u)
	}

	// Check if all nodes are connected using DFS
	visited := make(map[int]bool)
	dfs(0, graph, visited)

	return len(visited) == n
}

func dfs(node int, graph map[int][]int, visited map[int]bool) {
	visited[node] = true
	for _, neighbor := range graph[node] {
		if !visited[neighbor] {
			dfs(neighbor, graph, visited)
		}
	}
}

// UnionFind structure for cycle detection
type UnionFind struct {
	parent []int
	rank   []int
}

// NewUnionFind creates a new Union-Find structure
func NewUnionFind(n int) *UnionFind {
	parent := make([]int, n)
	rank := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = i
		rank[i] = 1
	}
	return &UnionFind{parent: parent, rank: rank}
}

// Find returns the root of the set containing x
func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x]) // Path compression
	}
	return uf.parent[x]
}

// Union merges the sets containing x and y
// Returns false if they're already in the same set (cycle detected)
func (uf *UnionFind) Union(x, y int) bool {
	rootX := uf.Find(x)
	rootY := uf.Find(y)

	if rootX == rootY {
		return false // Already in same set - cycle detected
	}

	// Union by rank
	if uf.rank[rootX] > uf.rank[rootY] {
		uf.parent[rootY] = rootX
	} else if uf.rank[rootX] < uf.rank[rootY] {
		uf.parent[rootX] = rootY
	} else {
		uf.parent[rootY] = rootX
		uf.rank[rootX]++
	}

	return true
}

// ValidTreeUnionFind checks if the edges form a valid tree using Union-Find.
// Time: O(E * α(n)) ≈ O(E), Space: O(n)
func ValidTreeUnionFind(n int, edges [][]int) bool {
	// A tree with n nodes must have exactly n-1 edges
	if len(edges) != n-1 {
		return false
	}

	uf := NewUnionFind(n)

	for _, edge := range edges {
		u, v := edge[0], edge[1]
		// If union returns false, we found a cycle
		if !uf.Union(u, v) {
			return false
		}
	}

	// If we have n-1 edges and no cycles, it's a valid tree
	return true
}
